In [148]:
import requests as rq
import pandas as pd
import linecache 



## Gathering ##

As requested in the project details, I get the data from the following sources:

### csv source: WeRateDogs Twitter archive  ### 

In [149]:
archive_df = pd.read_csv('twitter-archive-enhanced.csv')

### tsv source: image predictions ###

In [150]:
img_predictions = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = rq.get(img_predictions)

In [151]:
r.status_code

200

In [152]:
with open('image-predictions.tsv', 'wb') as f:
        f.write(r.content)

In [153]:
img_predictions_df  = pd.read_csv('image-predictions.tsv', sep= '\t')

### json source: querying the Twitter API ###

Using the code facilitated for this project (`twitter_api.py`), I downloaded json data for the tweet IDs contained in source one above. Once in the text file, I used the read_json function of Pandas:

In [154]:
json_df = pd.read_json('tweet_json.txt', lines = True )

## Assesing ##
> Detect and document at least eight (8) quality issues 
 and two (2) tidiness issues in your wrangle_act.ipynb Jupyter Notebook.
 To meet specifications, the issues that satisfy the Project Motivation (see the Key Points header on the previous page) must be assessed.

First, I proceed to look at the essential information concerning the three dataframes I just created. I do this, by invoking `info()` and `shape` for each df.

In [155]:
archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [156]:
archive_df.text.head()

0    This is Phineas. He's a mystical boy. Only eve...
1    This is Tilly. She's just checking pup on you....
2    This is Archie. He is a rare Norwegian Pouncin...
3    This is Darla. She commenced a snooze mid meal...
4    This is Franklin. He would like you to stop ca...
Name: text, dtype: object

In [157]:
archive_df.shape

(2356, 17)

In [158]:
img_predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [159]:
img_predictions_df.shape

(2075, 12)

We see all entries in the predicitions data frame has a valid URL 
#TODO find a proper URL regex and pass it to count()

In [160]:
img_predictions_df.jpg_url.str.count('http').size

2075

In [161]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2331 non-null   datetime64[ns, UTC]
 1   id                             2331 non-null   int64              
 2   id_str                         2331 non-null   int64              
 3   full_text                      2331 non-null   object             
 4   truncated                      2331 non-null   bool               
 5   display_text_range             2331 non-null   object             
 6   entities                       2331 non-null   object             
 7   extended_entities              2059 non-null   object             
 8   source                         2331 non-null   object             
 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_st

In [162]:
json_df.shape

(2331, 32)

The information above leaves us that the image predictions data frame has the least information with 2075 rows, next we have the json one with 2351, and finally the archive with 2356.

### Visual assessment ####
With our dataframes loaded, we can do a visual assessment, to get familiar with the data:

In [163]:
archive_df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
250   845098359547420673                    NaN                  NaN   
1899  674670581682434048                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
153   862457590147678208                    NaN                  NaN   
449   819004803107983360                    NaN                  NaN   

                      timestamp  \
250   2017-03-24 02:22:04 +0000   
1899  2015-12-09 19:22:56 +0000   
1068  2016-06-08 02:41:38 +0000   
153   2017-05-11 00:01:27 +0000   
449   2017-01-11 02:15:36 +0000   

                                                 source  \
250   <a href="http://twitter.com/download/iphone" r...   
1899  <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
153   <a href="http://twitter.com/download/iphone" r...   
449   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
250   RT @dog_rates: This is Bungalo. She uses that ...         7.733088e+17   
1899  Meet Jeb &amp; Bush. Jeb is somehow stuck in t...                  NaN   
1068  After so many requests, this is Bretagne. She ...                  NaN   
153   This is Jersey. He likes to watch movies, but ...                  NaN   
449   This is Bo. He was a very good First Doggo. 14...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
250               4.196984e+09  2016-09-06 23:56:05 +0000   
1899                       NaN                        NaN   
1068                       NaN                        NaN   
153                        NaN                        NaN   
449                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
250   https://twitter.com/dog_rates/status/773308824...                12   
1899  https://twitter.com/dog_rates/status/674670581...                 9   
1068  https://twitter.com/dog_rates/status/740373189...                 9   
153   https://twitter.com/dog_rates/status/862457590...                13   
449   https://twitter.com/dog_rates/status/819004803...                14   

      rating_denominator     name  doggo floofer pupper puppo  
250                   10  Bungalo   None    None   None  None  
1899                  10      Jeb   None    None   None  None  
1068                  11     None   None    None   None  None  
153                   10   Jersey   None    None   None  None  
449                   10       Bo  doggo    None   None  None

In [164]:
display(archive_df.rating_numerator.value_counts(ascending=False))
display(archive_df.rating_denominator.value_counts(ascending=False))

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
0         2
15        2
75        2
420       2
182       1
204       1
143       1
121       1
99        1
20        1
45        1
27        1
17        1
24        1
26        1
44        1
50        1
60        1
80        1
84        1
88        1
1776      1
960       1
666       1
144       1
165       1
Name: rating_numerator, dtype: int64

10     2333
11        3
50        3
20        2
80        2
0         1
120       1
7         1
170       1
150       1
130       1
90        1
110       1
2         1
70        1
40        1
16        1
15        1
Name: rating_denominator, dtype: int64

In [165]:
display(archive_df.doggo.value_counts(),  archive_df.floofer.value_counts())

None     2259
doggo      97
Name: doggo, dtype: int64

None       2346
floofer      10
Name: floofer, dtype: int64

In [166]:
display(archive_df.pupper.value_counts(), archive_df.puppo.value_counts())

None      2099
pupper     257
Name: pupper, dtype: int64

None     2326
puppo      30
Name: puppo, dtype: int64

In [167]:
predicate = archive_df.retweeted_status_id.isnull() == False
archive_df[predicate][['retweeted_status_id','retweeted_status_user_id','timestamp','retweeted_status_timestamp']]

retweeted_status_id  retweeted_status_user_id  \
19           8.874740e+17              4.196984e+09   
32           8.860537e+17              1.960740e+07   
36           8.305833e+17              4.196984e+09   
68           8.780576e+17              4.196984e+09   
73           8.782815e+17              4.196984e+09   
...                   ...                       ...   
1023         6.678667e+17              4.196984e+09   
1043         6.671383e+17              4.196984e+09   
1242         7.119983e+17              7.832140e+05   
2259         6.675487e+17              4.296832e+09   
2260         6.675484e+17              4.296832e+09   

                      timestamp retweeted_status_timestamp  
19    2017-07-21 01:02:36 +0000  2017-07-19 00:47:34 +0000  
32    2017-07-15 02:45:48 +0000  2017-07-15 02:44:07 +0000  
36    2017-07-13 01:35:06 +0000  2017-02-12 01:04:29 +0000  
68    2017-06-26 00:13:58 +0000  2017-06-23 01:10:23 +0000  
73    2017-06-24 00:09:53 +0000  2017-06-23 16:00:04 +0000  
...                         ...                        ...  
1023  2016-06-25 01:52:36 +0000  2015-11-21 00:46:50 +0000  
1043  2016-06-17 16:01:16 +0000  2015-11-19 00:32:12 +0000  
1242  2016-03-21 19:31:59 +0000  2016-03-21 19:29:52 +0000  
2259  2015-11-20 03:51:52 +0000  2015-11-20 03:43:06 +0000  
2260  2015-11-20 03:51:47 +0000  2015-11-20 03:41:59 +0000  

[181 rows x 4 columns]

In [168]:
img_predictions_df.sample(5)

tweet_id                                          jpg_url  \
1423  772117678702071809  https://pbs.twimg.com/media/Crcc7pqXEAAM5O2.jpg   
672   683142553609318400  https://pbs.twimg.com/media/CXsChyjW8AQJ16C.jpg   
1618  802572683846291456  https://pbs.twimg.com/media/CyNPmJgXcAECPuB.jpg   
357   672609152938721280  https://pbs.twimg.com/media/CVWWdKLWEAEnSk7.jpg   
1801  831939777352105988  https://pbs.twimg.com/media/C4uk0EWWQAAaZm1.jpg   

      img_num                  p1   p1_conf  p1_dog                  p2  \
1423        1  Labrador_retriever  0.217821    True              beagle   
672         1            Leonberg  0.605851    True                chow   
1618        1    golden_retriever  0.610171    True  Labrador_retriever   
357         1           microwave  0.981946   False          rotisserie   
1801        1          Pomeranian  0.153862    True              marmot   

       p2_conf  p2_dog                p3   p3_conf  p3_dog  
1423  0.157677    True  golden_retriever  0.127726    True  
672   0.183470    True   German_shepherd  0.079662    True  
1618  0.173252    True    cocker_spaniel  0.163257    True  
357   0.007472   False        television  0.005881   False  
1801  0.091234   False          grey_fox  0.090644   False

In [169]:
img_predictions_df[img_predictions_df['p1_dog'] == False].sample(5)

tweet_id                                            jpg_url  \
843   695064344191721472  https://pbs.twimg.com/ext_tw_video_thumb/69506...   
1416  771136648247640064    https://pbs.twimg.com/media/CrOgsIBWYAA8Dtb.jpg   
761   688894073864884227    https://pbs.twimg.com/media/CY9xf1dUAAE4XLc.jpg   
253   670727704916926465    https://pbs.twimg.com/media/CU7nSZEW4AA6r5u.jpg   
835   694183373896572928    https://pbs.twimg.com/media/CaI8Fn0WAAIrFJN.jpg   

      img_num           p1   p1_conf  p1_dog                p2   p2_conf  \
843         1    seat_belt  0.522211   False        sunglasses  0.077552   
1416        1      bathtub  0.368660   False  golden_retriever  0.297402   
761         1          hog  0.669996   False        guinea_pig  0.077347   
253         1  wood_rabbit  0.368562   False             tabby  0.309675   
835         1        teddy  0.441499   False          Pekinese  0.080870   

      p2_dog            p3   p3_conf  p3_dog  
843    False     ice_lolly  0.051774   False  
1416    True           tub  0.201711   False  
761    False       hamster  0.062398   False  
253    False  Egyptian_cat  0.154914   False  
835     True      Shih-Tzu  0.072099    True

In [170]:
img_predictions_df.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [171]:
json_df.sample(5)

created_at                  id              id_str  \
430  2017-01-11 02:57:27+00:00  819015337530290176  819015337530290176   
858  2016-08-02 18:14:06+00:00  760539183865880579  760539183865880576   
2242 2015-11-20 02:08:22+00:00  667524857454854144  667524857454854144   
1657 2015-12-29 23:32:35+00:00  681981167097122816  681981167097122816   
344  2017-02-10 16:53:37+00:00  830097400375152640  830097400375152640   

                                              full_text  truncated  \
430   RT @dog_rates: This is Bo. He was a very good ...      False   
858   This is Atlas. Swinging is his passion. 12/10 ...      False   
2242  Another topnotch dog. His name is Big Jumpy Ra...      False   
1657  This is Jimothy. He's a Trinidad Poliwhirl. Fa...      False   
344   Meet Samson. He's absolute fluffy perfection. ...      False   

     display_text_range                                           entities  \
430            [0, 122]  {'hashtags': [], 'symbols': [], 'user_mentions...   
858             [0, 64]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2242           [0, 136]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1657           [0, 136]  {'hashtags': [], 'symbols': [], 'user_mentions...   
344            [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
430   {'media': [{'id': 819004789207928832, 'id_str'...   
858   {'media': [{'id': 760539157622059009, 'id_str'...   
2242  {'media': [{'id': 667524856297218048, 'id_str'...   
1657  {'media': [{'id': 681981160138895360, 'id_str'...   
344   {'media': [{'id': 830097389742673920, 'id_str'...   

                                                 source  \
430   <a href="http://twitter.com/download/iphone" r...   
858   <a href="http://twitter.com/download/iphone" r...   
2242  <a href="http://twitter.com" rel="nofollow">Tw...   
1657  <a href="http://twitter.com/download/iphone" r...   
344   <a href="http://twitter.com/download/iphone" r...   

      in_reply_to_status_id  ...  favorited  retweeted  possibly_sensitive  \
430                     NaN  ...      False      False                 0.0   
858                     NaN  ...      False      False                 0.0   
2242                    NaN  ...      False      False                 0.0   
1657                    NaN  ...      False      False                 0.0   
344                     NaN  ...      False      False                 0.0   

     possibly_sensitive_appealable lang  \
430                            0.0   en   
858                            0.0   en   
2242                           0.0   en   
1657                           0.0   en   
344                            0.0   en   

                                       retweeted_status  quoted_status_id  \
430   {'created_at': 'Wed Jan 11 02:15:36 +0000 2017...               NaN   
858                                                 NaN               NaN   
2242                                                NaN               NaN   
1657                                                NaN               NaN   
344                                                 NaN               NaN   

     quoted_status_id_str  quoted_status_permalink  quoted_status  
430                   NaN                      NaN            NaN  
858                   NaN                      NaN            NaN  
2242                  NaN                      NaN            NaN  
1657                  NaN                      NaN            NaN  
344                   NaN                      NaN            NaN  

[5 rows x 32 columns]

In [172]:
#json_df[json_df.retweeted_status_id.isnull() == False]#.sample(3)
#json_df[.value_counts() ==]
display('favorite count stats')
display( json_df.favorite_count.describe())
display('retweet count stats')
display(json_df.retweet_count.describe())

'favorite count stats'

count      2331.000000
mean       7319.238953
std       11367.525636
min           0.000000
25%        1269.500000
50%        3174.000000
75%        8947.000000
max      150692.000000
Name: favorite_count, dtype: float64

'retweet count stats'

count     2331.000000
mean      2591.102531
std       4383.211411
min          1.000000
25%        525.500000
50%       1209.000000
75%       3000.000000
max      74376.000000
Name: retweet_count, dtype: float64

In [173]:
json_df.retweeted.value_counts()

False    2331
Name: retweeted, dtype: int64

In [174]:
json_df[['created_at', 'id']].head()

created_at                  id
0 2017-08-01 16:23:56+00:00  892420643555336193
1 2017-08-01 00:17:27+00:00  892177421306343426
2 2017-07-31 00:18:03+00:00  891815181378084864
3 2017-07-30 15:58:51+00:00  891689557279858688
4 2017-07-29 16:00:24+00:00  891327558926688256

In [175]:
archive_df[['timestamp', 'tweet_id']].head()

timestamp            tweet_id
0  2017-08-01 16:23:56 +0000  892420643555336193
1  2017-08-01 00:17:27 +0000  892177421306343426
2  2017-07-31 00:18:03 +0000  891815181378084864
3  2017-07-30 15:58:51 +0000  891689557279858688
4  2017-07-29 16:00:24 +0000  891327558926688256

In [176]:
json_df.full_text.head()

0    This is Phineas. He's a mystical boy. Only eve...
1    This is Tilly. She's just checking pup on you....
2    This is Archie. He is a rare Norwegian Pouncin...
3    This is Darla. She commenced a snooze mid meal...
4    This is Franklin. He would like you to stop ca...
Name: full_text, dtype: object

### Issues observed ###
From the (visual)assesment above, I can observe the following issues:

For the **archive dataset**:
* the following columns have missing values
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object

* the actual rating for the dogs is missing. According to the context of the project, it should be obtained from : rating_numerator / rating_denominator

* I interpret records with values in ['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'] as retweets, thus these entries won't be part of this project, as per the context given (sic."**You only want original ratings (no retweets) that have images**." )

* If I understand well, a dog can be either a doggo, a floofer, a pupper, or a puppo. This means the columns `doggo	floofer	pupper	puppo` can be melted(?) into a "stage" one?

For the **image predictions dataset**:

* img_predictions_df.jpg_url.value_counts() yields some image URL appear more than once, is this an issue?

* what does column img_num mean? Is it relevant to the project?

* column 'p1_dog' decides whether the entry can be regarded as a dog, yes or no. If columns p1* is the algorithm's #1 prediction for the image in the tweet, what I observe is that the image cannot be correctly predicted since there might be a dog in the image, but often is not the main subject in the image. Therefore I would remove those entries where 'p1_dog' is set to False*, being the rationale that as per the documentation

For the **json dataset**:

* the tweet id column is called `id` only. I prefer to rename it to `tweet_id` for consistency with other dataframes

* the following columns have missing values

 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_str      77 non-null     float64            
 11  in_reply_to_user_id            77 non-null     float64            
 12  in_reply_to_user_id_str        77 non-null     float64            
 13  in_reply_to_screen_name        77 non-null     object    
 15  geo                            0 non-null      float64            
 16  coordinates                    0 non-null      float64            
 17  place                          1 non-null      object             
 18  contributors                   0 non-null      float64 
 24  possibly_sensitive             2197 non-null   float64            
 25  possibly_sensitive_appealable  2197 non-null   float64 
 27  retweeted_status               163 non-null    object             
 28  quoted_status_id               26 non-null     float64            
 29  quoted_status_id_str           26 non-null     float64            
 30  quoted_status_permalink        26 non-null     object             
 31  quoted_status                  24 non-null     object 
 
   Moreover, since the requested columns from this data source are just three, the dataset needs to be pruned to keep only the incumbent columns (i.e. `tweet ID, retweet count, and favorite count`)

## Cleaning ##

Once identified the issues with the sources above, I start off this cleaning section, doing a backup of the "original" dataframes:

> Cleaning includes merging individual pieces of data according to the rules of tidy data.

In [177]:
archive_clean_df = archive_df
images_clean_df  = img_predictions_df
json_clean_df = json_df

In [ ]:
remove non interesting columns from json dataframe

In [179]:
json_clean_df.drop(labels= ['created_at', 'id_str', 'truncated', 'display_text_range', 'entities',
       'extended_entities', 'source', 'in_reply_to_status_id',
       'in_reply_to_status_id_str', 'in_reply_to_user_id',
       'in_reply_to_user_id_str', 'in_reply_to_screen_name', 'user', 'geo',
       'coordinates', 'place', 'contributors', 'is_quote_status', 'favorited',
       'retweeted', 'possibly_sensitive', 'possibly_sensitive_appealable',
       'lang', 'retweeted_status', 'quoted_status_id', 'quoted_status_id_str',
       'quoted_status_permalink', 'quoted_status'], inplace= True, axis=1)

test:

In [181]:
json_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              2331 non-null   int64 
 1   full_text       2331 non-null   object
 2   retweet_count   2331 non-null   int64 
 3   favorite_count  2331 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 73.0+ KB


rename id for tweet_id

In [189]:
json_clean_df.rename(columns={"id": "tweet_id"}, inplace=True)

test:

In [190]:
json_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2331 non-null   int64 
 1   full_text       2331 non-null   object
 2   retweet_count   2331 non-null   int64 
 3   favorite_count  2331 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 73.0+ KB


## Storing ##
> the clean files have to be joined and eventually save to a twitter_archive_master.csv

In [194]:
#merge  archive_clean_df with json_clean_d
master_df = archive_clean_df.merge(json_clean_df, how = 'left', right_on='tweet_id', left_on= 'tweet_id')
    
#join images_clean_df to archive_clean_df


test

In [197]:
master_df.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  \
0                  10  Phineas  None    None   None  None   
1                  10    Tilly  None    None   None  None   

                                           full_text  retweet_count  \
0  This is Phineas. He's a mystical boy. Only eve...         7369.0   
1  This is Tilly. She's just checking pup on you....         5492.0   

   favorite_count  
0         35042.0  
1         30349.0

and then we save it to the requested filename>

In [196]:
master_df.to_csv( 'twitter_archive_master.csv', index=False)

## Analyzing ##
>  At least three (3) insights . Some insights to obtain are:

 1) what's the dog best rated in the dataset. IN case of a tie, what;s the most recent?
 
 2) what's the dog worst rated in the dataset. IN case of a tie, what;s the most recent?
 
 3) what's the dog with more favorite tweets counts?

## Visualizing ##

In [ ]:
# At least one (1) visualization must be produced.
